File: tests/type/v2/timeline_test.ts
/**
 * Timeline型定義のテスト（TDD Red Phase）
 *
 * Timeline/TimelineEvent型が正しく定義されていることを検証
 */

import { assertEquals, assertExists } from "@std/assert";

// インポートテスト（この時点では失敗する）
import type {
  EventCategory,
  EventImportance,
  Timeline,
  TimelineDetails,
  TimelineDetectionHints,
  TimelineEvent,
  TimelineScope,
  TimePoint,
} from "@storyteller/types/v2/timeline.ts";

Deno.test("Timeline型定義", async (t) => {
  await t.step("EventCategory型が7種類のリテラル型であること", () => {
    // EventCategory型のテスト
    const categories: EventCategory[] = [
      "plot_point",
      "character_event",
      "world_event",
      "backstory",
      "foreshadow",
      "climax",
      "resolution",
    ];
    assertEquals(categories.length, 7);
  });

  await t.step("EventImportance型が3種類のリテラル型であること", () => {
    // EventImportance型のテスト
    const importances: EventImportance[] = ["major", "minor", "background"];
    assertEquals(importances.length, 3);
  });

  await t.step("TimelineScope型が4種類のリテラル型であること", () => {
    // TimelineScope型のテスト
    const scopes: TimelineScope[] = ["story", "world", "character", "arc"];
    assertEquals(scopes.length, 4);
  });

  await t.step(
    "TimePoint型がorder（必須）とオプショナルフィールドを持つこと",
    () => {
      // 必須のorderのみ
      const minimalTimePoint: TimePoint = {
        order: 1,
      };
      assertExists(minimalTimePoint.order);

      // 全フィールド指定
      const fullTimePoint: TimePoint = {
        order: 1,
        label: "物語開始",
        date: "春の初め",
        chapter: "chapter_01",
      };
      assertExists(fullTimePoint.order);
      assertEquals(fullTimePoint.label, "物語開始");
      assertEquals(fullTimePoint.date, "春の初め");
      assertEquals(fullTimePoint.chapter, "chapter_01");
    },
  );

  await t.step("TimelineEvent型が必須フィールドを持つこと", () => {
    const event: TimelineEvent = {
      id: "ball_invitation",
      title: "舞踏会への招待状",
      category: "plot_point",
      time: { order: 1 },
      summary: "シンデレラが舞踏会の招待状を受け取る",
      characters: ["cinderella", "stepmother"],
      settings: ["cinderella_house"],
      chapters: ["chapter_01"],
    };

    assertExists(event.id);
    assertExists(event.title);
    assertExists(event.category);
    assertExists(event.time);
    assertExists(event.summary);
    assertExists(event.characters);
    assertExists(event.settings);
    assertExists(event.chapters);
  });

  await t.step(
    "TimelineEvent型のcausedBy, causesがstring[]型であること",
    () => {
      const event: TimelineEvent = {
        id: "ball_dance",
        title: "舞踏会での踊り",
        category: "plot_point",
        time: { order: 2 },
        summary: "シンデレラと王子が踊る",
        characters: ["cinderella", "prince"],
        settings: ["royal_palace"],
        chapters: ["chapter_02"],
        causedBy: ["ball_invitation"],
        causes: ["midnight_escape"],
      };

      assertExists(event.causedBy);
      assertEquals(event.causedBy, ["ball_invitation"]);
      assertExists(event.causes);
      assertEquals(event.causes, ["midnight_escape"]);
    },
  );

  await t.step("TimelineEvent型のオプショナルフィールドを持てること", () => {
    const event: TimelineEvent = {
      id: "full_event",
      title: "完全なイベント",
      category: "climax",
      time: { order: 10 },
      summary: "物語のクライマックス",
      characters: ["hero"],
      settings: ["final_stage"],
      chapters: ["chapter_final"],
      causedBy: ["previous_event"],
      causes: ["resolution"],
      importance: "major",
      endTime: { order: 11, label: "クライマックス終了" },
      displayNames: ["最終決戦"],
      details: {
        description: "詳細な説明",
      },
      detectionHints: {
        commonPatterns: ["最終決戦で", "クライマックスの"],
        excludePatterns: [],
        confidence: 0.9,
      },
    };

    assertExists(event.importance);
    assertEquals(event.importance, "major");
    assertExists(event.endTime);
    assertExists(event.displayNames);
    assertExists(event.details);
    assertExists(event.detectionHints);
  });

  await t.step("Timeline型が必須フィールドを持つこと", () => {
    const timeline: Timeline = {
      id: "main_story",
      name: "メインストーリー",
      scope: "story",
      summary: "シンデレラの物語の主要タイムライン",
      events: [],
    };

    assertExists(timeline.id);
    assertExists(timeline.name);
    assertExists(timeline.scope);
    assertExists(timeline.summary);
    assertExists(timeline.events);
  });

  await t.step("Timeline型のオプショナルフィールドを持てること", () => {
    const timeline: Timeline = {
      id: "hero_journey",
      name: "勇者の旅",
      scope: "character",
      summary: "勇者キャラクターの個人タイムライン",
      events: [],
      parentTimeline: "main_story",
      childTimelines: ["side_quest_1", "side_quest_2"],
      relatedCharacter: "hero",
      displayNames: ["勇者の冒険"],
      displayOptions: {
        showRelations: true,
        colorScheme: "character",
      },
      details: {
        background: "勇者の旅の背景説明",
      },
    };

    assertExists(timeline.parentTimeline);
    assertEquals(timeline.parentTimeline, "main_story");
    assertExists(timeline.childTimelines);
    assertEquals(timeline.childTimelines?.length, 2);
    assertExists(timeline.relatedCharacter);
    assertExists(timeline.displayNames);
    assertExists(timeline.displayOptions);
    assertExists(timeline.details);
  });

  await t.step("TimelineDetails型がファイル参照をサポートすること", () => {
    const details: TimelineDetails = {
      background: { file: "timelines/main_story/background.md" },
      notes: "インラインのメモ",
    };

    assertExists(details.background);
    assertExists(details.notes);
  });

  await t.step("TimelineDetectionHints型が検出ヒントを持つこと", () => {
    const hints: TimelineDetectionHints = {
      commonPatterns: ["メインストーリーで", "物語の"],
      excludePatterns: ["サイドストーリー"],
      confidence: 0.85,
    };

    assertEquals(hints.commonPatterns.length, 2);
    assertEquals(hints.excludePatterns.length, 1);
    assertEquals(hints.confidence, 0.85);
  });
});

File: internal/domain/timeline_test.go
package domain_test

import (
	"testing"

	"github.com/takets/street-storyteller/internal/domain"
)

// TestTimelineZeroValue verifies Timeline can be constructed with required
// fields and remains a plain value type (no constructor needed).
func TestTimelineZeroValue(t *testing.T) {
	tl := domain.Timeline{
		ID:      "main_story",
		Name:    "メインストーリー",
		Scope:   domain.TimelineScopeStory,
		Summary: "物語全体の時系列",
		Events:  []domain.TimelineEvent{},
	}

	if tl.ID != "main_story" {
		t.Errorf("ID = %q, want %q", tl.ID, "main_story")
	}
	if tl.Scope != domain.TimelineScopeStory {
		t.Errorf("Scope = %q, want %q", tl.Scope, domain.TimelineScopeStory)
	}
	if tl.Events == nil {
		t.Error("Events should be non-nil empty slice")
	}

	// Optional fields default to zero values.
	if tl.ParentTimeline != nil {
		t.Error("ParentTimeline should be nil by default")
	}
	if tl.ChildTimelines != nil {
		t.Error("ChildTimelines should be nil by default")
	}
	if tl.RelatedCharacter != nil {
		t.Error("RelatedCharacter should be nil by default")
	}
	if tl.DisplayNames != nil {
		t.Error("DisplayNames should be nil by default")
	}
	if tl.DetectionHints != nil {
		t.Error("DetectionHints should be nil by default")
	}
	if tl.DisplayOptions != nil {
		t.Error("DisplayOptions should be nil by default")
	}
	if tl.Details != nil {
		t.Error("Details should be nil by default")
	}
}

// TestTimelineScopeConstants pins the four scope literal values from
// src/type/v2/timeline.ts so refactors cannot silently rename them.
func TestTimelineScopeConstants(t *testing.T) {
	cases := []struct {
		got  domain.TimelineScope
		want string
	}{
		{domain.TimelineScopeStory, "story"},
		{domain.TimelineScopeWorld, "world"},
		{domain.TimelineScopeCharacter, "character"},
		{domain.TimelineScopeArc, "arc"},
	}
	for _, c := range cases {
		if string(c.got) != c.want {
			t.Errorf("TimelineScope %q, want %q", c.got, c.want)
		}
	}
}

// TestEventCategoryConstants pins all EventCategory literal values.
func TestEventCategoryConstants(t *testing.T) {
	cases := []struct {
		got  domain.EventCategory
		want string
	}{
		{domain.EventCategoryPlotPoint, "plot_point"},
		{domain.EventCategoryCharacterEvent, "character_event"},
		{domain.EventCategoryWorldEvent, "world_event"},
		{domain.EventCategoryBackstory, "backstory"},
		{domain.EventCategoryForeshadow, "foreshadow"},
		{domain.EventCategoryClimax, "climax"},
		{domain.EventCategoryResolution, "resolution"},
	}
	for _, c := range cases {
		if string(c.got) != c.want {
			t.Errorf("EventCategory %q, want %q", c.got, c.want)
		}
	}
}

// TestEventImportanceConstants pins all EventImportance literal values.
func TestEventImportanceConstants(t *testing.T) {
	cases := []struct {
		got  domain.EventImportance
		want string
	}{
		{domain.EventImportanceMajor, "major"},
		{domain.EventImportanceMinor, "minor"},
		{domain.EventImportanceBackground, "background"},
	}
	for _, c := range cases {
		if string(c.got) != c.want {
			t.Errorf("EventImportance %q, want %q", c.got, c.want)
		}
	}
}

// TestTimelineEventZeroValue covers required fields and verifies optional
// causality/importance fields default to nil.
func TestTimelineEventZeroValue(t *testing.T) {
	ev := domain.TimelineEvent{
		ID:       "event_001",
		Title:    "物語の始まり",
		Category: domain.EventCategoryPlotPoint,
		Time:     domain.TimePoint{Order: 1},
		Summary:  "主人公が旅立つ",
		// Cross-entity ID slices are required (non-nil).
		Characters: []string{"hero"},
		Settings:   []string{"royal_capital"},
		Chapters:   []string{"chapter_01"},
	}

	if ev.ID != "event_001" {
		t.Errorf("ID = %q, want %q", ev.ID, "event_001")
	}
	if ev.Category != domain.EventCategoryPlotPoint {
		t.Errorf("Category = %q, want %q", ev.Category, domain.EventCategoryPlotPoint)
	}
	if ev.Time.Order != 1 {
		t.Errorf("Time.Order = %d, want 1", ev.Time.Order)
	}
	if len(ev.Characters) != 1 || ev.Characters[0] != "hero" {
		t.Errorf("Characters = %v, want [hero]", ev.Characters)
	}

	// Optional causality + importance default to nil.
	if ev.CausedBy != nil {
		t.Error("CausedBy should be nil by default")
	}
	if ev.Causes != nil {
		t.Error("Causes should be nil by default")
	}
	if ev.Importance != nil {
		t.Error("Importance should be nil by default")
	}
	if ev.EndTime != nil {
		t.Error("EndTime should be nil by default")
	}
	if ev.DisplayNames != nil {
		t.Error("DisplayNames should be nil by default")
	}
	if ev.Details != nil {
		t.Error("Details should be nil by default")
	}
	if ev.DetectionHints != nil {
		t.Error("DetectionHints should be nil by default")
	}
	if ev.PhaseChanges != nil {
		t.Error("PhaseChanges should be nil by default")
	}
}

// TestTimelineEventCausality verifies CausedBy / Causes accept string ID
// slices (cross-event references stay as IDs, not pointers).
func TestTimelineEventCausality(t *testing.T) {
	importance := domain.EventImportanceMajor
	ev := domain.TimelineEvent{
		ID:         "event_002",
		Title:      "決戦",
		Category:   domain.EventCategoryClimax,
		Time:       domain.TimePoint{Order: 10},
		Summary:    "最終決戦",
		Characters: []string{"hero", "villain"},
		Settings:   []string{"final_arena"},
		Chapters:   []string{"chapter_10"},
		CausedBy:   []string{"event_001"},
		Causes:     []string{"event_003"},
		Importance: &importance,
	}

	if got := ev.CausedBy; len(got) != 1 || got[0] != "event_001" {
		t.Errorf("CausedBy = %v, want [event_001]", got)
	}
	if got := ev.Causes; len(got) != 1 || got[0] != "event_003" {
		t.Errorf("Causes = %v, want [event_003]", got)
	}
	if ev.Importance == nil || *ev.Importance != domain.EventImportanceMajor {
		t.Errorf("Importance = %v, want major", ev.Importance)
	}
}

// TestTimePointOptionalFields verifies optional label/date/chapter use
// pointers (TS optional scalar -> Go pointer).
func TestTimePointOptionalFields(t *testing.T) {
	label := "第1日目"
	date := "1024-03-15"
	chapter := "chapter_01"
	tp := domain.TimePoint{
		Order:   1,
		Label:   &label,
		Date:    &date,
		Chapter: &chapter,
	}

	if tp.Order != 1 {
		t.Errorf("Order = %d, want 1", tp.Order)
	}
	if tp.Label == nil || *tp.Label != "第1日目" {
		t.Errorf("Label = %v, want %q", tp.Label, label)
	}
	if tp.Date == nil || *tp.Date != "1024-03-15" {
		t.Errorf("Date = %v, want %q", tp.Date, date)
	}

	// Zero TimePoint should have all-nil optionals.
	zero := domain.TimePoint{Order: 0}
	if zero.Label != nil || zero.Date != nil || zero.Chapter != nil {
		t.Error("zero TimePoint optionals should all be nil")
	}
}

// TestTimelineDetectionHints verifies the LSP hint struct mirrors the TS
// shape and is attachable as a pointer to Timeline.
func TestTimelineDetectionHints(t *testing.T) {
	hints := &domain.TimelineDetectionHints{
		CommonPatterns:  []string{"main story"},
		ExcludePatterns: []string{},
		Confidence:      0.85,
	}
	tl := domain.Timeline{
		ID:             "main_story",
		Name:           "メイン",
		Scope:          domain.TimelineScopeStory,
		Summary:        "概要",
		Events:         []domain.TimelineEvent{},
		DetectionHints: hints,
	}

	if tl.DetectionHints == nil {
		t.Fatal("DetectionHints should not be nil")
	}
	if tl.DetectionHints.Confidence != 0.85 {
		t.Errorf("Confidence = %f, want 0.85", tl.DetectionHints.Confidence)
	}
}

// TestTimelineDetailsFileRefUnion verifies the string|{file:string} union
// is modelled with the shared StringOrFileRef helper (Wave-A2-pre集約)。
func TestTimelineDetailsFileRefUnion(t *testing.T) {
	d := &domain.TimelineDetails{
		Background: &domain.StringOrFileRef{Value: "短い背景説明"},
		Notes:      &domain.StringOrFileRef{File: "notes.md"},
	}

	if d.Background == nil || d.Background.Value != "短い背景説明" {
		t.Errorf("Background.Value = %v, want 短い背景説明", d.Background)
	}
	if d.Notes == nil || d.Notes.File != "notes.md" {
		t.Errorf("Notes.File = %v, want notes.md", d.Notes)
	}
	if !d.Notes.IsFile() {
		t.Errorf("Notes should be reported as file ref")
	}
}

// TestPhaseChangeInfo confirms event-driven phase transitions can reference
// character / phase IDs as plain strings.
func TestPhaseChangeInfo(t *testing.T) {
	from := "phase_initial"
	desc := "覚醒"
	pc := domain.PhaseChangeInfo{
		CharacterID: "hero",
		ToPhaseID:   "phase_awakened",
		FromPhaseID: &from,
		Description: &desc,
	}
	if pc.CharacterID != "hero" || pc.ToPhaseID != "phase_awakened" {
		t.Errorf("PhaseChangeInfo = %+v", pc)
	}
	if pc.FromPhaseID == nil || *pc.FromPhaseID != from {
		t.Errorf("FromPhaseID = %v, want %q", pc.FromPhaseID, from)
	}
}
